src/ecs.rs

```diff
@@ -0,0 +1,692 @@
+            .fields()
+            .iter()
+            .find_map(|f| {
+                if f.key() == "observe_only" {
+                    Some(f.value() == "true")
+                } else {
+                    None
+                }
+            })
+            .unwrap_or_default();
+
+        for field in event.fields().iter() {
+            match field.key() {
+                "accept" => {}
+                "restrict" => {}
+                "unknown" => {}
+                "score" => {
+                    let mut risk = ecs_event.risk.clone().unwrap_or_default();
+                    risk.calculated_risk_score =
+                        field.value().parse::<f64>().map_err(|_| fmt::Error)?;
+                    ecs_event.risk = Some(risk);
+                }
+                "outcome" => {
+                    let mut risk = ecs_event.risk.clone().unwrap_or_default();
+                    let unquoted_outcome =
+                        quoted_string::to_content::<TraceQuoteSpec>(field.value())
+                            .map_err(|_| fmt::Error)?;
+                    risk.calculated_level = unquoted_outcome.to_ascii_lowercase();
+                    ecs_event.risk = Some(risk);
+
+                    // TODO: in the future it may be possible for user-defined handling of outcome and may require changes here
+                    let mut event_meta = ecs_event.event.clone().unwrap_or_default();
+                    if unquoted_outcome.to_ascii_lowercase() == "restricted" && !observe_only {
+                        event_meta.type_ = Some(vec![String::from("denied")]);
+                    } else {
+                        event_meta.type_ = Some(vec![String::from("allowed")]);
+                    }
+                    ecs_event.event = Some(event_meta);
+                }
+                "tags" => {
+                    let unquoted_tags = quoted_string::to_content::<TraceQuoteSpec>(field.value())
+                        .map_err(|_| fmt::Error)?;
+                    let tags: Vec<String> = unquoted_tags
+                        .to_string()
+                        .split(',')
+                        .map(|s| s.trim().to_ascii_lowercase())
+                        .filter(|s| !s.is_empty())
+                        .collect();
+                    if !tags.is_empty() {
+                        ecs_event.tags = Some(tags);
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Parses unrecognized messages on a "best effort" basis. Not currently implemented.
+    fn parse_unknown_event(event: &Event, _ecs_event: &mut EcsEvent) -> fmt::Result {
+        for _field in event.fields().iter() {
+            // TODO: make best guesses? right now these should rarely happen at info and debug level and user can switch to forest if needed
+        }
+        Ok(())
+    }
+
+    /// Parses a [`Span`] into an [`EcsEvent`]. Barely implemented.
```

Comment:
Yes, it's intentional to only implement part of ECS. It's very likely that I'll expand on what's implemented, but Bulwark will never use the complete ECS vocabulary. It's likely that just ECS serialization alone could be extracted and made generically useful, but I'm not in a hurry to tackle a side-project. I think it might be possible to codegen most of it though?